### DISCUSSION TITLE
Client-side page transitions do not work for catch-all routes

### DISCUSSION THREAD

**Discussion by keyworks** [vercel/next.js#15922] [Help]
# Bug report

## Describe the bug

When pushing a route that is handled by a catch-all page, I expect the routing to happen client-side. Instead, the entire page reloads and server-side routing is used. 

For example, suppose I have defined the following page:
```
pages/
  [[...slug]].js
  page2.js
```
And suppose I have the following links defined somewhere:
```
<Link href="/page1">Page 1</Link>
<Link href="/page2">Page 2</Link>
```

When I click on `Page 1`, the page reloads as the routing happens server-side which is not ideal.
When I click on `Page 2`, the routing happens client-side which is ideal.

## To Reproduce

1. `git clone https://github.com/keyworks/next-catch-all-client-routing-issue.git`
2. `yarn dev` and visit `localhost:3000`
3. Click on `Page 1` link
4. Observe the entire page reloads (the page will flicker).

## Expected behavior

Page should not reload, but instead page transition should happen client side. 

## Screenshots

n/a

## System information

- OS: macOS
- Browser (if applies: Firefox 79.0
- Version of Next.js: 9.5.1
- Version of Node.js: 10.22.0

## Additional context

n/a

**Comment by AubreyHewes** [comment]
You should do:

`<Link href="/[...slug]" as="/page1">Page 1</Link>`

Which entails a page file as `[...slug].js`

OR

`<Link href="/[slug]" as="/page1">Page 1</Link>`

Which entails a page file as `[slug].js`

The difference between `[slug]` and `[...slug]` is described in the documentation..

Hard (known?) pages can be done via `href`

This is all in the docs ;-) x

**Comment by keyworks** [comment]
This to me feels like an inconsistency in the API hence the bug report. For example, when I want a `Link` to client-side transition to `page2.js`, I do not need to define both a `asPath` and `href` param. 

I would expect next client router to automatically know about catch-all page routes so that if I attempt to call `router.push('/page1')` or click on `<Link href="/page1"/>`, it would trigger a client-side page transition to `[...slug].js`. Instead the client router doesn't seem to know it exists so it defers to server routing and the whole document reloads.
